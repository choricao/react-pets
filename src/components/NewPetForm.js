import React, { Component } from 'react';
import PropTypes from 'prop-types';

class NewPetForm extends Component {

  constructor() {
    super();

    this.state = {
      name: '',
      age: 0,
      breed: '',
      about: '',
    }
  }

  onFieldChange = (event) => {
    const fieldName = event.target.name;
    const fieldValue =
      fieldName === 'age' ? parseInt(event.target.value) : event.target.value;
    const updateState = {};
    updateState[fieldName] = fieldValue;
    this.setState(updateState);
  }

  valid = () => {
    return this.state.name.length > 0 && this.state.age > 0;
  }

  clearForm = () => {
    this.setState({
      name: '',
      age: 0,
      breed: '',
      about: '',
    });
  }

  onFormSubmit = (event) => {
    event.preventDefault();

    if (this.valid()) {
      this.props.addPetCallback(this.state);

      this.clearForm();
    }
  }

  render() {
    return(
      <form onSubmit={this.onFormSubmit}>
        <div>
          <label htmlFor="name">Name: </label>
          <input
            name="name"
            value={this.state.name}
            onChange={this.onFieldChange}
            type="text" />
        </div>
        <div>
          <label htmlFor="age">Age: </label>
          <input
            name="age"
            value={this.state.age}
            onChange={this.onFieldChange}
            type="number" />
        </div>
        <div>
          <label htmlFor="breed">Breed: </label>
          <input
            name="breed"
            value={this.state.breed}
            onChange={this.onFieldChange}
            type="text" />
        </div>
        <div>
          <label htmlFor="about">About: </label>
          <textarea
            name="about"
            value={this.state.about}
            onChange={this.onFieldChange} />
        </div>
        <input type="submit" value="Add Pet" />
      </form>
    );
  }

}

NewPetForm.propTypes = {
  addPetCallback: PropTypes.func.isRequired,
}

export default NewPetForm;
