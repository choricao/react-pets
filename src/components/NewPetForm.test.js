import React from 'react';
import NewPetForm from './NewPetForm';
import { mount, shallow } from 'enzyme';

describe('NewPetForm', () => {
  test('that it matches an existing snapshot', () => {
    // First Mount the Component in the testing DOM
    // Arrange
    const wrapper = mount(<NewPetForm
      addPetCallback={() => {}}
      />);

    // Assert that it looks like the last snapshot
    expect(wrapper).toMatchSnapshot();

    // Remove the component from the DOM (save memory and prevent side effects).
    wrapper.unmount();
  });

  test('When a user enters a name in a text field the field is updated', () => {
    // Arrange
    const wrapper = shallow(<NewPetForm
      addPetCallback={() => {}}
      />);
    let nameField = wrapper.find('input[name="name"]');

    // Act
    nameField.simulate('change', {
      target: {
        name: 'name',
        value: 'Bob',
      }
    });
    // Force the onChange event
    wrapper.update();
    nameField = wrapper.find('input[name="name"]');

    // Assert
    expect(nameField.getElement().props.value).toEqual('Bob');
  });
});
