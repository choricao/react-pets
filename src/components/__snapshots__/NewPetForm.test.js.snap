// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPetForm that it matches an existing snapshot 1`] = `
<NewPetForm
  addPetCallback={[Function]}
>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="name"
      >
        Name: 
      </label>
      <input
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        htmlFor="age"
      >
        Age: 
      </label>
      <input
        name="age"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </div>
    <div>
      <label
        htmlFor="breed"
      >
        Breed: 
      </label>
      <input
        name="breed"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        htmlFor="about"
      >
        About: 
      </label>
      <textarea
        name="about"
        onChange={[Function]}
        value=""
      />
    </div>
    <input
      type="submit"
      value="Add Pet"
    />
  </form>
</NewPetForm>
`;
